src/analysis/transcendental/liouville.lean

```diff
@@ -0,0 +1,408 @@
+  { apply div_nonneg, norm_cast, assumption, norm_cast, linarith }
+end
+
+lemma abs_eval_rat_ge_one_div_denom_pow_n (f : polynomial ℤ) (a b : ℤ)
+  (b_pos : 0 < b) (not_root : (f.map (algebra_map ℤ ℝ)).eval (a/b) ≠ 0) :
+  1/(b : ℝ)^f.nat_degree ≤ abs ((f.map (algebra_map ℤ ℝ)).eval (a/b)) :=
+have b_nonzero : b ≠ 0, from ne_of_gt b_pos,
+have pos : 0 < 1 / (b : ℝ) ^ f.nat_degree, from
+begin
+  refine div_pos (show 0 < (1 : ℝ), from by linarith) (pow_pos _ _),
+  norm_cast, exact b_pos,
+end,
+have eq₁ : _, from
+  calc ((f.map (algebra_map ℤ ℝ)).eval (a/b))
+      = ∑ i in f.support, f.coeff i * (a / b) ^ i
+      : begin
+          conv_lhs { rw [eval_map, as_sum_support f, eval₂_finset_sum] },
+          simp only [coeff_map, eval₂_X_pow, div_pow, eval₂_mul, eval₂_C],
+          congr
+        end
+  ... = ∑ i in f.support, f.coeff i * a^i / b^i : sum_congr rfl (λ _ _, by rw [div_pow, mul_div_assoc])
+  ... = ∑ i in f.support, f.coeff i * a^i * b^(-i:ℤ) : sum_congr rfl (λ _ _, by {ring, simp} )
+  ... = ∑ i in f.support, f.coeff i * a^i * b^(-i+ f.nat_degree - f.nat_degree : ℤ)
+      : sum_congr rfl (λ _ _, by { simp only [add_sub_cancel] })
+  ... = ∑ i in f.support, f.coeff i * a^i * b^(-i + f.nat_degree:ℤ)/b^f.nat_degree
+      : sum_congr rfl (λ _ _, by { rw [fpow_sub, fpow_coe_nat, mul_div_assoc], norm_cast, assumption })
+  ... = ∑ i in f.support, f.coeff i * a^i * b^(f.nat_degree - i:ℤ)/b^f.nat_degree
+      : sum_congr rfl (λ _ _, by rw [add_comm _ ↑f.nat_degree, sub_eq_add_neg])
+  ... = ∑ i in f.support, f.coeff i * a^i * b^(f.nat_degree - i)/b^f.nat_degree
+      : sum_congr rfl (λ _ h, by { rw [←int.sub_nat_nat_eq_coe, int.sub_nat_nat_of_ge (le_nat_degree_of_mem_support h), int.of_nat_eq_coe], refl })
+  ... = (∑ i in f.support, f.coeff i * a^i * b^(f.nat_degree - i))/b^f.nat_degree : by rw [sum_div]
+  ... = 1/b^f.nat_degree * (∑ i in f.support, f.coeff i*a^i*b^(f.nat_degree - i)) : by ring,
+have cast_eq : ∑ i in f.support, (f.coeff i : ℝ) * a ^ i * b ^ (f.nat_degree - i)
+            = ↑(∑ i in f.support, f.coeff i * a ^ i * b ^ (f.nat_degree - i)), from
+begin
+  induction f.support using finset.induction_on with a s H ih,
+  { simp only [int.cast_zero, sum_empty] },
+  { rw [sum_insert H, sum_insert H, ih],
+    simp only [int.cast_pow, int.cast_add, int.cast_mul] }
+end,
+begin
+  rw [eq₁, cast_eq, abs_mul, abs_of_pos pos, le_mul_iff_one_le_right pos],
+  rw [eq₁, mul_ne_zero_iff, cast_eq] at not_root,
+  norm_cast at ⊢ not_root,
+  suffices : 0 < abs (∑ (i : ℕ) in f.support, f.coeff i * a ^ i * b ^ (f.nat_degree - i)),
+  { omega },
+  { rw abs_pos_iff,
+    exact not_root.2 }
+end
+
+lemma exists_pos_real_of_irrational_root (α : ℝ) (hα : irrational α) (f : polynomial ℤ)
+  (f_deg : 1 < f.nat_degree) (α_root : (f.map (algebra_map ℤ ℝ)).eval α = 0) :
+  ∃ A : ℝ, 0 < A ∧ ∀ a b : ℤ, 0 < b → (A / b ^ (f.nat_degree)) < abs(α - a / b)  :=
+have f_nonzero : f ≠ 0, from ne_zero_of_nat_degree_gt f_deg,
```

Comment:
Why not just put the `have` statements inside the `begin ... end` block?
Term-mode is nice, but I would avoid writing multiple `begin ... end`s if you can also put everything in one block.